File: CSharpApp/MainApp/UiRibbon/Tabs/TabCreator.cs
using System;
using System.Collections.Generic;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.Windows;
using MainApp.UiRibbon.Panels;
using Microsoft.VisualBasic;

namespace MainApp.UiRibbon.Tabs
{
    public class TabCreator
    {
        public static void AddTabs(Document doc)
        {
            var appRibbonTab = AppTab.CreateAppTab();
            PanelCreator.AddAllPanels(appRibbonTab);
            var allRibbonTabs = new List<RibbonTab>(new RibbonTab[] { appRibbonTab });
            AddTabs(doc, allRibbonTabs);
        }

        private static void AddTabs(Document doc, List<RibbonTab> ribbonTabs)
        {
            doc.Editor.WriteMessage(Constants.vbLf + "Adding tabs...");
            foreach (RibbonTab ribbonTab in ribbonTabs)
                AddTab(doc, ribbonTab);
        }

        private static void AddTab(Document doc, RibbonTab ribbonTab)
        {
            doc.Editor.WriteMessage(Constants.vbLf + "Add tab...");
            var ribbonControl = ComponentManager.Ribbon;
            if (ribbonControl != null & ribbonTab != null)
            {
                if (ribbonTab.Name == null)
                {
                    throw new Exception("Ribbon Tab does not have a name. Please add a name to the ribbon tab.");
                }
                string tabName = ribbonTab.Name;
                doc.Editor.WriteMessage(Constants.vbLf + tabName);
                var doesTabAlreadyExist = ribbonControl.FindTab(tabName);
                if (doesTabAlreadyExist != null)
                {
                    ribbonControl.Tabs.Remove(doesTabAlreadyExist);
                }
                ribbonControl.Tabs.Add(ribbonTab);
                ribbonTab.IsActive = true;
            }
        }
    }
}

File: VbNetApp/MainApp/UiRibbon/Tabs/TabCreator.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.Windows
Imports MainApp.UiRibbon.Panels

Namespace UiRibbon.Tabs
    Public Class TabCreator
        Public Shared Sub AddTabs(doc As Document)
            Dim appRibbonTab As RibbonTab = AppTab.CreateAppTab()
            PanelCreator.AddAllPanels(appRibbonTab)
            Dim allRibbonTabs As New List(Of RibbonTab)(New RibbonTab() {appRibbonTab})
            AddTabs(doc, allRibbonTabs)
        End Sub

        Private Shared Sub AddTabs(doc As Document, ribbonTabs As List(Of RibbonTab))
            doc.Editor.WriteMessage(vbLf & "Adding tabs...")
            For Each ribbonTab As RibbonTab In ribbonTabs
                AddTab(doc, ribbonTab)
            Next
        End Sub

        Private Shared Sub AddTab(doc As Document, ribbonTab As RibbonTab)
            doc.Editor.WriteMessage(vbLf & "Add tab...")
            Dim ribbonControl As RibbonControl = ComponentManager.Ribbon
            If (ribbonControl IsNot Nothing And
                ribbonTab IsNot Nothing) Then
                If (ribbonTab.Name Is Nothing) Then
                    Throw New Exception("Ribbon Tab does not have a name. Please add a name to the ribbon tab.")
                End If
                Dim tabName As String = ribbonTab.Name
                doc.Editor.WriteMessage(vbLf & tabName)
                Dim doesTabAlreadyExist As RibbonTab = ribbonControl.FindTab(tabName)
                If (doesTabAlreadyExist IsNot Nothing) Then
                    ribbonControl.Tabs.Remove(doesTabAlreadyExist)
                End If
                ribbonControl.Tabs.Add(ribbonTab)
                ribbonTab.IsActive = True
            End If
        End Sub
    End Class
End Namespace
